imports/server/activityTracking.ts

```diff
@@ -0,0 +1,123 @@
+}
+
+function serializeActivityKey({ ts, hunt, puzzle }: ActivityKey) {
+  return `${ts.getTime()}-${hunt}-${puzzle}`;
+}
+
+function deserializeActivityKey(key: string) {
+  const [ts, hunt, puzzle] = key.split('-');
+  if (!ts || !hunt || !puzzle) {
+    throw new Error(`Invalid activity key: ${key}`);
+  }
+  return { ts: new Date(parseInt(ts, 10)), hunt, puzzle };
+}
+
+async function fetchAndDeleteRecentActivity(
+  cutoff: Date,
+  session: InstanceType<typeof MongoInternals.NpmModules.mongodb.module.ClientSession>,
+) {
+  if (!session) {
+    throw new Error('Must be called with a session');
+  }
+
+  const usersByKeyAndType = new Map<string, Map<RecentActivityType['type'], Set<string>>>();
+  while (true) {
+    /* eslint-disable no-await-in-loop */
+    const recent = await RecentActivities.rawCollection().findOneAndDelete(
+      { ts: { $lt: cutoff } } as any,
+      { session }
+    );
+    if (!recent.value) {
+      break;
+    }
+
+    const key = serializeActivityKey(recent.value);
+    const usersByType = usersByKeyAndType.get(key) ?? new Map();
+    const users = usersByType.get(recent.value.type) ?? new Set();
+    users.add(recent.value.user);
+    usersByType.set(recent.value.type, users);
+    usersByKeyAndType.set(key, usersByType);
+    /* eslint-enable no-await-in-loop */
+  }
+
+  return usersByKeyAndType;
+}
+
+async function consolidateRecentActivity() {
+  const ageForConsolidation = ACTIVITY_GRANULARITY * 2;
+  const cutoff = new Date(Date.now() - ageForConsolidation);
+
+  // By using a transaction, we avoid races with other servers, since we only
+  // count records that we delete (and if we fail to update the consolidated
+  // counters, the deletions will be rolled back).
+  //
+  // However: note that using the raw collection bypasses our schema
+  // enforcement, so we need to make sure of compliance
+  const { client } = MongoInternals.defaultRemoteCollectionDriver().mongo;
+  const session = client.startSession();
+  await session.withTransaction(async () => {
+    const usersByKeyAndType = await fetchAndDeleteRecentActivity(
+      cutoff,
+      session,
+    );
+
+    await [...usersByKeyAndType.entries()].reduce(async (p, [key, usersByType]) => {
+      await p;
+
+      const matcher = deserializeActivityKey(key);
+      const increments = new Map<'total' | `components.${RecentActivityType['type']}`, number>();
+      const allUsers = new Set<string>();
+      [...usersByType.entries()].forEach(([type, users]) => {
+        const incrementKey = `components.${type}` as const;
+        increments.set(incrementKey, (increments.get(incrementKey) ?? 0) + users.size);
+        users.forEach((user) => allUsers.add(user));
+      });
+      increments.set('total', allUsers.size);
+
+      await ConsolidatedActivities.rawCollection().updateOne(
+        { ...matcher },
+        {
+          $setOnInsert: {
+            _id: Random.id(),
+          },
+          $inc: Object.fromEntries(increments),
```

Comment:
In normal operation, we don't expect to see these records getting updated, only inserted, right?  Because if we did, then we'd potentially double-count user ids.

Reply:
Oh hmm...there's some risk of concurrency - multiple servers could potentially be running consolidation simultaneously (and I assumed that would be fine because of the transactions). I may have to think a little more about how to do this correctly.

Reply:
This is actually safe - I sort of assumed that MongoDB would do the dumbest thing possible and allow dirty reads within a transaction (i.e. the equivalent of `READ UNCOMMITTED`). However, MongoDB does seem to do something more intelligent here. From [https://www.mongodb.com/docs/manual/core/transactions/#transactions-and-atomicity](Transactions and Atomicity):

> Until a transaction commits, the data changes made in the transaction are not visible outside the transaction.

So even if two servers are running this process concurrently, they won't see the other process's deletes, the transactions will conflict, and everything will be fine.